packages/gpl-core/src/badge.ts

```diff
@@ -0,0 +1,314 @@
+import { SDK } from ".";
+import { gql } from "graphql-request";
+import * as anchor from "@project-serum/anchor";
+import randomBytes from "randombytes";
+
+export interface GraphQLBadge {
+  address: string;
+  issuer: string;
+  holder: string;
+  update_authority: string;
+  schema: string;
+  metadata_uri: string;
+  refreshed_at?: Date;
+  slot_created_at?: Date;
+  slot_updated_at?: Date;
+  created_at?: Date;
+}
+
+export interface GraphQLIssuer {
+  address: string;
+  authority: string;
+  verified: boolean;
+  refreshed_at?: Date;
+  slot_created_at?: Date;
+  slot_updated_at?: Date;
+  created_at?: Date;
+}
+
+export class Badge {
+  private readonly sdk: SDK;
+
+  constructor(sdk: SDK) {
+    this.sdk = sdk;
+  }
+
+  public async getBadge(userAccount: anchor.web3.PublicKey) {
+    return await this.sdk.program.account.badge.fetch(userAccount);
+  }
+
+  public async getOrCreateBadge(
+    metadataUri: string,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    holder: anchor.web3.PublicKey,
+    updateAuthority: anchor.web3.PublicKey,
+    authority: anchor.web3.PublicKey,
+  ) {
+    const [badgePDA, _] = anchor.web3.PublicKey.findProgramAddressSync(
+      [
+        Buffer.from("badge"),
+        issuer.toBuffer(),
+        schema.toBuffer(),
+        holder.toBuffer(),
+      ],
+      this.sdk.program.programId
+    );
+
+    try {
+      await this.getBadge(badgePDA);
+    } catch (err) {
+      await (await this.createBadge(metadataUri, issuer, schema, holder, updateAuthority, authority)).instructionMethodBuilder.rpc();
+    }
+    return badgePDA;
+  }
+
+  public async createBadge(
+    metadataUri: string,
+    issuer: anchor.web3.PublicKey,
+    schema: anchor.web3.PublicKey,
+    holderProfile: anchor.web3.PublicKey,
+    updateAuthority: anchor.web3.PublicKey,
+    authority: anchor.web3.PublicKey) {
+    const instructionMethodBuilder = this.sdk.program.methods
+      .createBadge(metadataUri)
+      .accounts({
+        issuer,
+        schema,
+        holder: holderProfile,
+        updateAuthority,
+        authority,
+      });
+    const pubKeys = await instructionMethodBuilder.pubkeys();
```

Comment:
This is no longer necessary, you can just use [rpcAndKeys](https://github.com/coral-xyz/anchor/blob/master/ts/packages/anchor/src/program/namespace/methods.ts#LL222C1-L231C4).

Reply:
Not sure if there is a way to do it without making the rpc call

Reply:
this method intentionally avoids making rpc calls. It is designed for fetching instructions or transactions that can be grouped depending on the use case.